scheds/rust/scx_bpfland/src/bpf/main.bpf.c

```diff
@@ -527,50 +592,57 @@ static s32 pick_idle_cpu(struct task_struct *p, s32 prev_cpu, u64 wake_flags, bo
 	idle_cpumask = __COMPAT_scx_bpf_get_idle_cpumask_node(node);
 
 	/*
-	 * If the current task is waking up another task and releasing the CPU
-	 * (WAKE_SYNC), attempt to migrate the wakee on the same CPU as the
-	 * waker.
+	 * Check whether the waker and wakee are in the same LLC and
+	 * evaluate a migration if the LLC is not fully saturated.
 	 */
-	cpu = bpf_get_smp_processor_id();
-	if (is_wake_sync(p, current, cpu, prev_cpu, wake_flags)) {
-		const struct cpumask *curr_l3_domain;
-		struct cpu_ctx *cctx;
-		bool share_llc;
+	share_llc = l3_mask && bpf_cpumask_test_cpu(this_cpu, l3_mask) &&
+		    !is_llc_busy(idle_cpumask, this_cpu);
 
+	/*
+	 * In case of a wakeup from a different CPU, attempt to migrate the
+	 * wakee to the waker's CPU.
+	 *
+	 * We don't want to do cross-LLC migrations here, because they
+	 * can be really expensive in certain architectures (e.g., recent
+	 * AMD CPUs).
+	 */
+	if (share_llc && is_wake_remote(prev_cpu, this_cpu, wake_flags)) {
 		/*
-		 * Determine waker CPU scheduling domain.
+		 * If @prev_cpu is in a full-idle SMT core, try to reuse
+		 * the same CPU and avoid any migration, since there is no
+		 * guarantee that the cache hot data from a waker
+		 * (interrupt or task) is more important than cache hot
+		 * data on @prev_cpu.
 		 */
-		cctx = try_lookup_cpu_ctx(cpu);
-		if (!cctx) {
-			cpu = -EINVAL;
+		if ((!smt_enabled || bpf_cpumask_test_cpu(prev_cpu, idle_smtmask)) &&
+		    scx_bpf_test_and_clear_cpu_idle(prev_cpu)) {
+			cpu = prev_cpu;
+			*is_idle = true;
 			goto out_put_cpumask;
 		}
 
-		curr_l3_domain = cast_mask(cctx->l3_cpumask);
-		if (!curr_l3_domain)
-			curr_l3_domain = primary;
-
 		/*
-		 * If both the waker and wakee share the same L3 cache keep
-		 * using the same CPU if possible.
+		 * If this is a wakeup from an idle CPU, then it's an IRQ
+		 * wakeup and we can attempt to migrate the task to the
+		 * IRQ CPU.
 		 */
-		share_llc = bpf_cpumask_test_cpu(prev_cpu, curr_l3_domain);
-		if (share_llc &&
-		    scx_bpf_test_and_clear_cpu_idle(prev_cpu)) {
-			cpu = prev_cpu;
+		if ((!smt_enabled || bpf_cpumask_test_cpu(this_cpu, idle_smtmask)) &&
+		    scx_bpf_test_and_clear_cpu_idle(this_cpu)) {
+			cpu = this_cpu;
 			*is_idle = true;
 			goto out_put_cpumask;
 		}
```

Comment:
Might be worth trying here if nothing is found to return a CPU in the LLC anyways? This is the part that I tried testing and couldn't get good results on.

Reply:
Yeah that's the culprit, it's more or less what fair.c is doing, but in our case it seems to introduce more overhead than benefits. Maybe it's good in some special cases, but it seems a bit too situational.

I think it makes sense to return any CPU in the LLC, but that means removing this chunk of code and fallback to the regular path of finding an idle CPU. So... basically we can drop the last commit and we get better performance in general.